### PR TITLE
infra: implement blue/green deployments for production (step 2)

### PR DESCRIPTION
**Pull request by guifry**
This PR implements [blue/green deployments](https://martinfowler.com/bliki/BlueGreenDeployment.html) for the production API.

## Blue Green resources

The distinction will be made on Cloud Run revisions. Anything else in the infrastructure will remained unchanged.
Using Cloud Run revisions is handy because Cloud Run supports manual traffic allocation out of the box.

## Logic Flow

In the Pulumi script, right before creating the Cloud Run instance:

1. Fetches the list of revisions in the Cloud Run service
2. Gets the latest revision colour
3. Infers from it the next colour to be deployed next
4. Sets up the traffic configuration for both the new revision and the previous one.

At the moment, **any new production deployment will be allocated 5% traffic**, whilst the previous revision will be set at 95%. We will have to manually adjust it progressively in the console, as we gain confidence in the new revision.

## Setting and fetching a revision's colour

The revisions's colour is specified in its annotations. A revision annotation is a list of key/value pairs present in the revision YAML template (see screenshot).

Initially, I had planned to do it via the revision tags, as it's easier to visualise in the console. However, it turns out that the GCP SDK doesn't allow us to query a list of tags for our revisions, making it difficult to keep track of the previously deployed revisions and infer the new traffic configuration. This is a well known feature ask from the community.

Setting up annotation and fetching it with the SDK is more straightforward.

I am still using tags, but it's just to indicate if the revision is blue or green (see screenshot). We still need the annotation for the script to know what colour was last deployed.

Screenshot 2023-11-22 at 14 54 11

image

### CODE REVIEW

File: infra/utils.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    sorted_revs = sorted(revisions.revisions, key=lambda r: r.create_time, reverse=True)
+
+    latest_rev = sorted_revs[0]
+    try:
+        col_annotation = latest_rev.annotations["production-color"]
+        latest_color = ProductionColor(col_annotation)
+    except Exception as e:
+        raise Exception(
+            f"Unknown color annotation on latest revision: {col_annotation}"
+        ) from e
```
Comment (andyhd) [inline infra/utils.py]: Question: Can we be more specific than `except Exception`? And can we raise a more specific Exception too? I think `ValueError` fits here.

File: infra/utils.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    service_name = "production-consent-api"
+    parent = f"projects/{project_id}/locations/{region}/services/{service_name}"
+
+    revisions = revclient.list_revisions(parent=parent)
+
+    sorted_revs = sorted(revisions.revisions, key=lambda r: r.create_time, reverse=True)
+
+    latest_rev = sorted_revs[0]
+    try:
+        col_annotation = latest_rev.annotations["production-color"]
```
Comment (andyhd) [inline infra/utils.py:26]: Question, non-blocking: I'm assuming for the first Blue/Green deployment, we'll manually add an annotation to the current revision?

**Review by andyhd — CHANGES_REQUESTED** [review]
Praise: This is awesome! 🥳 It's amazing that you can do this without much code. Good to go, apart from a greedy `try ... except`

File: infra/utils.py (hunk 1 above)
Comment (guifry) [reply infra/utils.py]: Good shout, indeed `ValueError` is the one here

File: infra/utils.py (hunk 2 above)
Comment (guifry) [reply infra/utils.py:26]: I've already done something similar and tested a couple of follow-up deployments: https://console.cloud.google.com/run/detail/europe-west2/production-consent-api/revisions?project=sde-consent-api

Comment (andyhd) [reply infra/utils.py:26]: Of course, the screenshots are in the PR 🤦  Nice one!

**Review by andyhd — APPROVED** [review]
:shipit:

**Pull request merged**